packages/cache/src/internal/tar.ts

```diff
@@ -49,19 +54,28 @@ function getWorkingDirectory(): string {
 }
 
 // Common function for extractTar and listTar to get the compression method
-function getCompressionProgram(compressionMethod: CompressionMethod): string[] {
+async function getCompressionProgram(compressionMethod: CompressionMethod): Promise<string[]> {
   // -d: Decompress.
   // unzstd is equivalent to 'zstd -d'
   // --long=#: Enables long distance matching with # bits. Maximum is 30 (1GB) on 32-bit OS and 31 (2GB) on 64-bit.
   // Using 30 here because we also support 32-bit self-hosted runners.
+  const tarPath = await getTarPath([])
+  const BSD_TAR_ZSTD = IS_WINDOWS && tarPath === SystemTarPathOnWindows
   switch (compressionMethod) {
     case CompressionMethod.Zstd:
+      if (BSD_TAR_ZSTD) {
+        return ['-a'] // auto-detect compression
```

Comment:
This is done to make it future proof?

Reply:
Do we not have to pipe here?

Reply:
Future proof as well as simpler to reason about as the ordering of args might get brittle with different platforms.

Reply:
Won't we still need pipe?

Reply:
We should not need to. Unless this also exhibits the hanging issue.

Reply:
Have to check it

Reply:
I think no. `-a` seems to be useful for compression not decompression. I got the below error while running it:
`Option -a is not permitted in mode -x`. Workflow run: https://github.com/bbq-beets/test-zstd/actions/runs/3512158872/jobs/5883582602
freebsd docs also stats the same: https://www.freebsd.org/cgi/man.cgi?query=bsdtar&sektion=1&format=htm